rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -26,75 +26,103 @@ use super::{select_neighbors, select_neighbors_heuristic, HNSW};
+    pub m: usize,
+
+    /// max number of connections for each element per layers.
+    pub m_max: usize,
+
+    /// size of the dynamic list for the candidates
+    pub ef_construction: usize,
+
+    /// whether extend candidates while selecting neighbors
+    pub extend_candidates: bool,
+
+    /// log base used for assigning random level
+    pub log_base: f32,
+
+    /// whether select neighbors heuristic
+    pub use_select_heuristic: bool,
+}
+
+impl Default for HnswBuildParams {
+    fn default() -> Self {
+        Self {
+            max_level: 7,
+            m: 20,
+            m_max: 30,
+            ef_construction: 25,
+            extend_candidates: false,
+            log_base: 10.0,
+            use_select_heuristic: true,
+        }
+    }
+}
+
 /// Build a HNSW graph.
 ///
 /// Currently, the HNSW graph is fully built in memory.
 ///
 /// During the build, the graph is built layer by layer.
 ///
 /// Each node in the graph has a global ID which is the index on the base layer.
+#[derive(Clone)]
 pub struct HNSWBuilder {
-    /// max level of
-    max_level: u16,
-
-    /// number of connections to establish while inserting new element
-    m: usize,
-
-    /// max number of connections for each element per layers.
-    m_max: usize,
-
-    /// Size of the dynamic list for the candidates
-    ef_construction: usize,
+    params: HnswBuildParams,
 
     /// Vector storage for the graph.
     vectors: Arc<InMemoryVectorStorage>,
 
     levels: Vec<GraphBuilder>,
 
     entry_point: u32,
-
-    extend_candidates: bool,
-
-    log_base: f32,
-
-    use_select_heuristic: bool,
 }
 
 impl HNSWBuilder {
     /// Create a new [`HNSWBuilder`] with in memory vector storage.
     pub fn new(vectors: Arc<InMemoryVectorStorage>) -> Self {
         Self {
-            max_level: 8,
-            m: 30,
-            m_max: 64,
-            ef_construction: 100,
+            params: HnswBuildParams::default(),
+            vectors,
+            levels: vec![],
+            entry_point: 0,
+        }
+    }
+
+    /// Create a new [`HNSWBuilder`] with prepared params and in memory vector storage.
+    pub fn with_params(params: HnswBuildParams, vectors: Arc<InMemoryVectorStorage>) -> Self {
+        Self {
+            params: params,
             vectors,
             levels: vec![],
             entry_point: 0,
-            extend_candidates: false,
-            log_base: 10.0,
-            use_select_heuristic: true,
         }
     }
 
     /// The maximum level of the graph.
     /// The default value is `8`.
     pub fn max_level(mut self, max_level: u16) -> Self {
-        self.max_level = max_level;
+        self.params.max_level = max_level;
         self
     }
 
     /// The number of connections to establish while inserting new element
     /// The default value is `30`.
     pub fn num_edges(mut self, m: usize) -> Self {
-        self.m = m;
+        self.params.m = m;
```

Comment:
nit: since everything is `pub` in the `params`. we could just do `pub fn params(mut self, params)` to have less boiler plate